presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java

```diff
@@ -331,15 +350,74 @@ private boolean checkTablePermission(ConnectorTransactionHandle transaction, Ide
+        Set<HivePrivilege> privilegeSet = listApplicableTablePrivileges(
+                metastore,
+                tableName.getSchemaName(),
+                tableName.getTableName(),
+                new PrestoPrincipal(USER, identity.getUser()))
+                .stream()
                 .map(HivePrivilegeInfo::getHivePrivilege)
-                .collect(Collectors.toSet());
+                .collect(toSet());
         return privilegeSet.containsAll(ImmutableSet.copyOf(requiredPrivileges));
     }
 
     private boolean isAdmin(ConnectorTransactionHandle transaction, Identity identity)
     {
         SemiTransactionalHiveMetastore metastore = metastoreProvider.apply(((HiveTransactionHandle) transaction));
-        return metastore.getRoles(identity.getUser()).contains(ADMIN_ROLE_NAME);
+        return listApplicableRoles(metastore, new PrestoPrincipal(USER, identity.getUser())).contains(ADMIN_ROLE_NAME);
+    }
+
+    private static Set<HivePrivilegeInfo> getDatabasePrivileges(SemiTransactionalHiveMetastore metastore, String user, String databaseName)
+    {
+        Set<HivePrivilegeInfo> privileges = new HashSet<>();
+        if (isDatabaseOwner(metastore, user, databaseName)) {
+            privileges.add(new HivePrivilegeInfo(OWNERSHIP, true));
+        }
+        return privileges;
+    }
+
+    private static boolean isDatabaseOwner(SemiTransactionalHiveMetastore metastore, String user, String databaseName)
+    {
+        // all users are "owners" of the default database
+        if (DEFAULT_DATABASE_NAME.equalsIgnoreCase(databaseName)) {
+            return true;
+        }
+
+        Optional<Database> databaseMetadata = metastore.getDatabase(databaseName);
+        if (!databaseMetadata.isPresent()) {
+            return false;
+        }
+
+        Database database = databaseMetadata.get();
+
+        // a database can be owned by a user or role
+        if (database.getOwnerType() == USER && user.equals(database.getOwnerName())) {
+            return true;
+        }
+        if (database.getOwnerType() == ROLE && listApplicableRoles(metastore, new PrestoPrincipal(USER, user)).contains(database.getOwnerName())) {
+            return true;
+        }
+        return false;
+    }
+
+    private static Set<String> listApplicableRoles(SemiTransactionalHiveMetastore metastore, PrestoPrincipal principal)
+    {
+        return MetastoreUtil.listApplicableRoles(principal, metastore::listRoleGrants)
+                .stream()
+                .map(RoleGrant::getRoleName)
+                .collect(toSet());
+    }
+
+    private static Set<HivePrivilegeInfo> listApplicableTablePrivileges(SemiTransactionalHiveMetastore metastore, String databaseName, String tableName, PrestoPrincipal principal)
+    {
+        Set<String> applicableRoles = listApplicableRoles(metastore, principal);
+        List<PrestoPrincipal> principals = new ArrayList<>();
```

Comment:
This is just a local variable. I'm not sure if it worth extra code to change it to builder.